### PR TITLE
NDRS-{1052,1053}: Prevent deploy replays. (Ports #1216, #1218, #1219, #1223 to dev.)

### PR DESCRIPTION
**Pull request by afck**
This is a port of https://github.com/CasperLabs/casper-node/pull/1216, https://github.com/CasperLabs/casper-node/pull/1218, https://github.com/CasperLabs/casper-node/pull/1219, https://github.com/CasperLabs/casper-node/pull/1223 to the `dev` branch.

* Instead of persisting the `BlockProposerDeploySets` to LMDB and loading in on startup, we recreate that state from the collection of finalized deploys. This fixes a bug where that set would be incorrectly initialized for joining/restarting/upgrading nodes.
* Blocks are considered invalid if they contain the same deploy twice, or if they contain a deploy that was already present in any of their ancestors.

https://casperlabs.atlassian.net/browse/NDRS-1052
https://casperlabs.atlassian.net/browse/NDRS-1053

### CODE REVIEW

File: node/src/components/block_validator.rs [hunk 1]
```diff
@@ -149,11 +149,20 @@ where
...
+                let deploy_count = block_deploys.len();
+                // Collect the deploys in a set; this also deduplicates them.
+                let block_deploys: HashSet<_> = block_deploys
                     .iter()
                     .map(|deploy_hash| **deploy_hash)
-                    .collect::<HashSet<_>>();
+                    .collect();
+                if block_deploys.len() != deploy_count {
+                    info!(
+                        deploys = ?block.deploys(), ?sender,
```
Comment (goral09) [inline node/src/components/block_validator.rs:161]: really, we're logging 2600 deploy hashes? or even whole bodies?

**Review by goral09 — APPROVED** [review]

File: node/src/components/block_validator.rs (hunk 1 above)
Comment (afck) [reply node/src/components/block_validator.rs:161]: If we receive a block with an attempted replay attack? Why not?

Comment (afck) [reply node/src/components/block_validator.rs:161]: On the other hand: The deploy hashes won't actually tell us much; we should rather log the block proposer maybe?

Comment (goral09) [reply node/src/components/block_validator.rs:161]: My question is _why_?

Comment (goral09) [reply node/src/components/block_validator.rs:161]: I would say that we should log the block proposer and replayed hashes. This shouldn't happen often so we can afford finding duplicates.

Comment (afck) [reply node/src/components/block_validator.rs:161]: Turns out we only have a `BlockLike` there, not a `Block`, so no proposer etc.
Maybe we should move this check into consensus after all?

Comment (afck) [reply node/src/components/block_validator.rs]: As discussed, I changed the log to only list the duplicated deploys.

**Pull request merged**